File: swift/Sources/FieldUIPlatform/FrameScheduler.swift
import Foundation
import FieldUICore

// MARK: - Phases

/// The six-phase frame loop — discover→read→compute→state→write→render.
/// Phase order is a hard invariant: handlers always run in this sequence.
public enum Phase: Int, CaseIterable, CustomStringConvertible, Sendable {
    case discover, read, compute, state, write, render

    public var description: String {
        switch self {
        case .discover: "discover"
        case .read:     "read"
        case .compute:  "compute"
        case .state:    "state"
        case .write:    "write"
        case .render:   "render"
        }
    }
}

/// Phases in which reading geometry is legal.
public let readPhases: Set<Phase> = [.discover, .read]

// MARK: - Frame context

public struct FrameContext {
    public let now: TimeInterval
    public let volume: FieldVolume
    public let phase: Phase
    public let frame: Int
}

public typealias PhaseHandler = (FrameContext) -> Void

// MARK: - Violations

public struct PhaseViolation {
    public let phase: Phase
    public let op: String
    public let allowed: [Phase]
    public let frame: Int
    public var message: String {
        "\"\(op)\" ran in the \(phase) phase; allowed only in: \(allowed.map(\.description).joined(separator: ", "))"
    }
}

public struct FrameReport {
    public let frame: Int
    public let now: TimeInterval
    public let ran: [Phase]
    public let violations: [PhaseViolation]
}

// MARK: - FrameScheduler

/// One shared loop with explicit, ordered phases so the registries never fight each other.
///
/// Every frame walks the same six phases in the same order:
///   discover → read → compute → state → write → render
///
/// The scheduler enforces the discipline: a geometry read requested during the write phase
/// is a violation. In strict mode it throws; otherwise it records and continues.
public final class FrameScheduler {
    /// A registered handler tagged with a stable token, so an unsubscribe identifies the original
    /// handler even after earlier removals shift the array (a captured index would go stale).
    private struct Subscription { let id: Int; let handler: PhaseHandler }
    private var handlers: [Phase: [Subscription]] = {
        var d = [Phase: [Subscription]]()
        for p in Phase.allCases { d[p] = [] }
        return d
    }()
    private var nextToken = 0

    private let strict: Bool
    private var recorded: [PhaseViolation] = []
    private(set) var current: Phase? = nil
    private(set) var frame: Int = 0

    public init(strict: Bool = false) {
        self.strict = strict
    }

    // MARK: Registration

    /// Register a handler for a phase. Returns an unsubscribe closure.
    @discardableResult
    public func on(_ phase: Phase, _ handler: @escaping PhaseHandler) -> () -> Void {
        let id = nextToken
        nextToken += 1
        handlers[phase]!.append(Subscription(id: id, handler: handler))
        return { [weak self] in
            self?.handlers[phase]?.removeAll { $0.id == id }
        }
    }

    // MARK: Phase guard

    /// Called by registries before reading geometry. Off-phase reads are violations.
    public func assertReadPhase(op: String) {
        guard let phase = current else { return } // outside a frame — allowed
        guard readPhases.contains(phase) else {
            let v = PhaseViolation(phase: phase, op: op, allowed: Array(readPhases), frame: frame)
            if strict { fatalError("[FieldUI/Platform] \(v.message)") }
            recorded.append(v)
            return
        }
    }

    public func readGuard() -> (String) -> Void {
        { [weak self] op in self?.assertReadPhase(op: op) }
    }

    // MARK: Violations

    public func violations() -> [PhaseViolation] { recorded }
    public func clearViolations() { recorded.removeAll() }

    // MARK: Frame execution

    /// Run one full six-phase frame. Returns the per-frame report.
    @discardableResult
    public func runFrame(now: TimeInterval = 0, volume: FieldVolume = FieldVolume(width: 0, height: 0)) -> FrameReport {
        let startViolations = recorded.count
        var ran: [Phase] = []

        for phase in Phase.allCases {
            let list = handlers[phase] ?? []
            guard !list.isEmpty else { continue }
            current = phase
            ran.append(phase)
            let ctx = FrameContext(now: now, volume: volume, phase: phase, frame: frame)
            for sub in list { sub.handler(ctx) }
        }

        current = nil
        let report = FrameReport(
            frame: frame,
            now: now,
            ran: ran,
            violations: Array(recorded[startViolations...])
        )
        frame += 1
        return report
    }
}

File: android/fundamental-platform/src/main/kotlin/com/fundamental/platform/FrameScheduler.kt
package com.fundamental.platform

import com.fundamental.core.engine.FieldVolume

/**
 * The six-phase frame loop — ported from FrameScheduler.swift line-for-line.
 * discover → read → compute → state → write → render. Phase order is a hard invariant.
 */
enum class Phase { DISCOVER, READ, COMPUTE, STATE, WRITE, RENDER;

    override fun toString(): String = name.lowercase()
}

/** Phases in which reading geometry is legal. */
val readPhases: Set<Phase> = setOf(Phase.DISCOVER, Phase.READ)

data class FrameContext(
    val now: Double,
    val volume: FieldVolume,
    val phase: Phase,
    val frame: Int,
)

typealias PhaseHandler = (FrameContext) -> Unit

data class PhaseViolation(
    val phase: Phase,
    val op: String,
    val allowed: List<Phase>,
    val frame: Int,
) {
    val message: String
        get() = "\"$op\" ran in the $phase phase; allowed only in: ${allowed.joinToString(", ")}"
}

data class FrameReport(
    val frame: Int,
    val now: Double,
    val ran: List<Phase>,
    val violations: List<PhaseViolation>,
)

/**
 * One shared loop with explicit, ordered phases so the registries never fight each other.
 * A geometry read requested during the write phase is a violation: in strict mode it throws,
 * otherwise it is recorded and the frame continues.
 */
class FrameScheduler(private val strict: Boolean = false) {

    private class Subscription(val id: Int, val handler: PhaseHandler)
    private val handlers: MutableMap<Phase, MutableList<Subscription>> =
        Phase.entries.associateWith { mutableListOf<Subscription>() }.toMutableMap()
    private var nextToken = 0

    private val recorded = ArrayList<PhaseViolation>()
    var current: Phase? = null
        private set
    var frame: Int = 0
        private set

    /** Register a handler for a phase. Returns an unsubscribe closure. */
    fun on(phase: Phase, handler: PhaseHandler): () -> Unit {
        val id = nextToken++
        handlers.getValue(phase).add(Subscription(id, handler))
        return { handlers[phase]?.removeAll { it.id == id } }
    }

    /** Called by registries before reading geometry. Off-phase reads are violations. */
    fun assertReadPhase(op: String) {
        val phase = current ?: return // outside a frame — allowed
        if (phase !in readPhases) {
            val v = PhaseViolation(phase, op, readPhases.toList(), frame)
            if (strict) error("[Fundamental/Platform] ${v.message}")
            recorded.add(v)
        }
    }

    fun readGuard(): (String) -> Unit = { op -> assertReadPhase(op) }

    fun violations(): List<PhaseViolation> = recorded.toList()
    fun clearViolations() = recorded.clear()

    /** Run one full six-phase frame. Returns the per-frame report. */
    fun runFrame(now: Double = 0.0, volume: FieldVolume = FieldVolume(0f, 0f)): FrameReport {
        val startViolations = recorded.size
        val ran = ArrayList<Phase>()
        for (phase in Phase.entries) {
            val list = handlers[phase] ?: continue
            if (list.isEmpty()) continue
            current = phase
            ran.add(phase)
            val ctx = FrameContext(now, volume, phase, frame)
            // snapshot to tolerate unsubscribe during iteration
            for (sub in list.toList()) sub.handler(ctx)
        }
        current = null
        val report = FrameReport(frame, now, ran, recorded.subList(startViolations, recorded.size).toList())
        frame += 1
        return report
    }
}
